File: Solutions/198-house-robber/house-robber.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  # If there are no houses, you cannot rob anything
  return 0 if nums.empty?

  # If there is only one house, rob it
  return nums[0] if nums.length == 1

  # prev2 will store the maximum money rob-able up to the house before the previous house
  prev2 = 0
  # prev1 stores the maximum money rob-able up to the previous house
  prev1 = nums[0]

  # Iterate over the remaining houses starting from the second house
  nums[1..-1].each do |money|
    # If we rob this house, add its money to prev2 (since we can't rob prev1)
    # If we skip this house, the max money stays prev1
    current = [prev1, prev2 + money].max

    # Shift prev1 and prev2 for next iteration
    prev2 = prev1
    prev1 = current
  end

  # At the end, prev1 contains the maximum money we can rob
  prev1
end

File: Solutions/198-house-robber/house-robber.scala
object Solution {
    def rob(nums: Array[Int]): Int = {

        // Edge case: if no houses
        if (nums.isEmpty) return 0

        // Edge case: if only one house
        if (nums.length == 1) return nums(0)

        // prev1 = dp[i-1] → max money till previous house
        // prev2 = dp[i-2] → max money till house before previous
        var prev1 = 0
        var prev2 = 0

        // Iterate through each house
        for (money <- nums) {

            // Option 1: skip this house → prev1
            // Option 2: rob this house → prev2 + money
            val current = Math.max(prev1, prev2 + money)

            // Move forward:
            // prev2 becomes old prev1
            prev2 = prev1

            // prev1 becomes current result
            prev1 = current
        }

        // Final result
        prev1
    }
}